chirp/drivers/tk11.py

```diff
@@ -0,0 +1,3297 @@
+    """Validate nullable float"""
+    if value is None or value == "":
+        return ""
+
+    try:
+        float_val = float(value)
+    except ValueError as ve:
+        raise ValueError("Value is not float or empty") from ve
+
+    if float_val < FM_RANGE[0] or float_val > FM_RANGE[1]:
+        raise ValueError(f"Value is not in range {FM_RANGE}")
+
+    return str(float_val)
+
+
+def do_download(radio):
+    """This is your download function"""
+    # Get the serial port connection
+    serial = radio.pipe
+
+    status = chirp_common.Status()
+    status.cur = 0
+    status.msg = "Connecting to radio"
+    radio.status_fn(status)
+
+    connection_request_message = encapsulate_message(
+        ConnectionRequest.to_bytes()
+    )
+
+    try:
+        serial.write(connection_request_message)
+    except Exception as se:
+        raise errors.RadioError(
+            "Error with serial, failed to write connection request message"
+        ) from se
+
+    connection_serial_response = read_serial(serial, 1000)
+
+    if connection_serial_response is None:
+        LOG.error("Invalid radio model, connection response not received")
+        raise errors.RadioError(
+            "Invalid radio model, connection response not received"
+        )
+
+    connection_response = ConnectionResponse.from_bytes(
+        connection_serial_response
+    )
+
+    if connection_response.message_type != MessageType.CONNECT_RESPONSE.value:
+        LOG.error("Invalid radio connection response")
+        raise errors.RadioError("Invalid radio connection response")
+
+    LOG.info(
+        "Connected to radio - firmware version : %s",
+        connection_response.version
+    )
+
+    radio.firmware_version = connection_response.version
+
+    status.msg = "Reading channels usage from radio"
+    radio.status_fn(status)
+
+    memory = bytearray()
+
+    # channels
+    channels_usage = bytearray()
+    for index in range(2):
+        channels_usage.extend(
+            read_memory(
+                serial,
+                MemoryAddress.CHANNELS_USAGE + (index * MAX_CHUNK_SIZE),
+                MAX_CHUNK_SIZE
+            )
+        )
+
+    channels_usage = channels_usage[:CHANNELS_COUNT]
+
+    status.msg = "Reading channels from radio"
+    radio.status_fn(status)
+
+    radio.edited_channels = []
+
+    channel_struct_size = 64
+    channels = bytearray()
+    for i_channel in range(CHANNELS_COUNT):
+        if channels_usage[i_channel] != 0xFF:
+            status.msg = f"Reading channel #{i_channel + 1} from radio"
+            radio.status_fn(status)
+            channel_response = read_memory(
+                serial,
+                i_channel * channel_struct_size,
+                channel_struct_size
+            )
+            channels.extend(channel_response)
+            radio.edited_channels.append(i_channel)
+
+    channels.extend(
+        [0] * (
+            (CHANNELS_COUNT - int(len(channels) / channel_struct_size))
```

Comment:
instead of `int()` here, just use `//`.

Also, this very spaced out zero-bad list comprehension is not very easy to read, IMHO.